File: algos/range_queries/PrefixSum/solnJava/dwivediprashant/Solution1.java
// 1. Problem Statement -------------------------------------------------
/*
Given an integer array nums, handle multiple range sum queries.
Each query asks for the sum of elements between indices left and right (inclusive).
The array is immutable, meaning its values never change.
*/

// 2. Approach -----------------------------------------------------------
/*
We use the Prefix Sum technique.

A prefix sum array is constructed where:
prefixSum[i] = nums[0] + nums[1] + ... + nums[i]

Using this array:
- If left == 0, the range sum is prefixSum[right]
- Otherwise, the range sum is prefixSum[right] - prefixSum[left - 1]

The prefix sum array is built once and reused for all queries.
*/

// 3. Complexity ---------------------------------------------------------
/*
Preprocessing Time: O(n)
Query Time: O(1)
Space Complexity: O(n)
*/

// Compiling Code in Java (Solution1.java) -------------------------------

package CodeCraft.algos.range_queries.PrefixSum.solnJava.dwivediprashant;

public class Solution1 {

    private int[] prefixSum;

    public Solution1(int[] nums) {
        int n = nums.length;
        prefixSum = new int[n];
        if (n == 0) return;
        prefixSum[0] = nums[0];
        for (int i = 1; i < n; i++) {
            prefixSum[i] = prefixSum[i - 1] + nums[i];
        }
    }

    public int sumRange(int left, int right) {
        if (left == 0) {
            return prefixSum[right];
        }
        return prefixSum[right] - prefixSum[left - 1];
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 4, 5};
        Solution1 obj = new Solution1(nums);
        System.out.println(obj.sumRange(1, 3));
    }
}

File: algos/range_queries/PrefixSum/soln/Jatin/Solution1.cpp
/*
Problem:
LeetCode 303 - Range Sum Query: Immutable

Given an integer array nums, handle multiple queries:
find the sum of elements between indices L and R inclusive.

Approach:
Build a prefix sum array where:
prefix[i] = sum of nums[0...i]

Then:
sum(L, R) = prefix[R] - prefix[L-1]   (if L > 0)

Time Complexity:
Preprocessing: O(N)
Each Query: O(1)

Space Complexity: O(N)

Example:
Input: nums = [-2,0,3,-5,2,-1], query(0,2)
Output: 1
*/

#include <iostream>
#include <vector>
using namespace std;

class NumArray {
    vector<long long> prefix;
public:
    NumArray(vector<int>& nums) {
        int n = nums.size();
        prefix.resize(n);
        if (n > 0) {
            prefix[0] = nums[0];
            for(int i = 1; i < n; i++)
                prefix[i] = prefix[i - 1] + nums[i];
        }
    }

    long long sumRange(int L, int R) {
        if (L == 0) return prefix[R];
        return prefix[R] - prefix[L - 1];
    }
};

int main() {
    vector<int> nums = {-2, 0, 3, -5, 2, -1};
    NumArray obj(nums);
    cout << obj.sumRange(0, 2) << endl; // 1
    cout << obj.sumRange(2, 5) << endl; // -1
    return 0;
}
